### PR TITLE
[REVIEW] Implement count_nonzero_mask in libgdf

### PR DESCRIPTION
**Pull request by jrhemstad**
This PR will implement a function to count the number of valid bits in a validity bitmask.

### CODE REVIEW

File: src/validops.cu [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+  auto mask_bit_counter = [masks, num_masks, num_rows] __device__ (int index)
+  {
+
+    constexpr int BITS_PER_MASK = sizeof(gdf_valid_type) * 8;
+
+    gdf_valid_type count = 0;
+    gdf_valid_type mask = masks[index];
+    if(index < num_masks - 1)
+    {
+      // FIXME: If gdf_valid_type was 32 bits, we could use __popc
```
Comment (harrism) [inline src/validops.cu]: Can't you process in groups of 4 bytes (except the last one) and use __popc?

File: src/validops.cu [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+      while( mask > 0 )
```
Comment (harrism) [inline src/validops.cu]: Since mask has type unsigned char, it will always be positive, so I don't see how this loop can exit.

Comment (jrhemstad) [reply src/validops.cu]: It exits when `0 == mask`.

File: src/validops.cu [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+        count += mask & gdf_valid_type(1);
+        mask >>= 1;
+      }
+    }
+    return count;
+  };
+
+  thrust::device_vector<gdf_valid_type> mask_counts(num_masks);
+
+  // FIXME: We could use transform_reduce if we didn't have to
```
[the remark this answers is not in the document]
Comment (jrhemstad) [reply src/validops.cu]: Yeah, that's what I thought of shortly after I had to leave for the day. I was thinking you'd copy the "remainder masks" to host in a separate stream, and then do a `transform_reduce` on the number of masks that are a multiple of 4 bytes. You could then count the number of bits in the non-multiple of 4 masks (remainder masks) on the host and add them to the result from `transform_reduce`.

File: src/validops.cu (hunk 1 above)
Comment (jrhemstad) [reply src/validops.cu]: Indeed, see updated implementation.

File: src/validops.cu [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+    gdf_valid_type current_mask = masks[i];
+
+    while(current_mask > 0)
+    {
+      current_mask &= (current_mask-1) ;
+      count++;
+    }
+  }
+
+  constexpr int BITS_PER_MASK = sizeof(gdf_valid_type) * 8;
```
Comment (harrism) [inline src/validops.cu]: Shouldn't `8` here be `GDF_VALID_BITSIZE`?

**Comment by scopatz** [comment]
LGTM Other than the BITSIZE issue

**Pull request merged**